util/pkg/vfs/swiftfs.go

```diff
@@ -0,0 +1,453 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vfs
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-ini/ini"
+	"github.com/golang/glog"
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	swiftcontainer "github.com/gophercloud/gophercloud/openstack/objectstorage/v1/containers"
+	swiftobject "github.com/gophercloud/gophercloud/openstack/objectstorage/v1/objects"
+	"github.com/gophercloud/gophercloud/pagination"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/kops/util/pkg/hashing"
+)
+
+func NewSwiftClient() (*gophercloud.ServiceClient, error) {
+	config := OpenstackConfig{}
+
+	authOption, err := config.GetCredential()
+	if err != nil {
+		return nil, err
+	}
+	provider, err := openstack.AuthenticatedClient(authOption)
+	if err != nil {
+		return nil, fmt.Errorf("error building openstack authenticated client: %v", err)
+	}
+	endpointOpt, err := config.GetServiceConfig("Swift")
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := openstack.NewObjectStorageV1(provider, endpointOpt)
+	if err != nil {
+		return nil, fmt.Errorf("error building swift client: %v", err)
+	}
+	return client, nil
+}
+
+type OpenstackConfig struct {
+}
+
+func (_ OpenstackConfig) filename() (string, error) {
+	name := os.Getenv("OPENSTACK_CREDENTIAL_FILE")
+	if name != "" {
+		return name, nil
```

Comment:
As a tip, you might want to (for example) `glog.V(2).Infof("using openstack config from credentials found in OPENSTACK_CREDENTIAL_FILE: %s", name)` etc

The idea being that if someone reports "I don't know where openstack is getting its configuration from" or "My openstack is not picking up the right credentials" we can ask them to run with `--v=2` and get a hint as to why.

But just a suggestion for the future :-)

Reply:
good suggestion! will update it!